tests/ports.py

```diff
@@ -0,0 +1,119 @@
+"""
+Test Port mappings work on running non-jupyter workflows
+"""
+import subprocess
+import urllib3
+import time
+import os
+import tempfile
+import signal
+import random
+
+
+def read_port_mapping_response(host, port, protocol = None):
+    """
+    Deploy container and test if port mappings work as expected
+
+    Args:
+        host: the host interface to bind to.
+        port: the random host port to bind to
+        protocol: the protocol to use valid values /tcp or /udp
+    """
+    builddir = os.path.dirname(__file__)
+    port_protocol = '8000'
+    if protocol:
+        port_protocol += protocol
+    host_port = port
+    if host:
+        host_port = host + ':' + port
+    else:
+        host = 'localhost'
+    with tempfile.TemporaryDirectory() as tmpdir:
+        username = os.getlogin()
+
+        proc = subprocess.Popen(['repo2docker',
+                                 '-p',
+                                 host_port + ':' + port_protocol,
+                                 '-v', '{}:/home'.format(tmpdir),
+                                 '--user-id', str(os.geteuid()),
+                                 '--user-name', username,
+                                 '.',
+                                 '/bin/bash', '-c', 'echo \'hi\' > /home/ts && python -m http.server 8000'],
+                                 cwd=builddir + "/../",
+                                 stderr=subprocess.STDOUT)
+        try:
+            while not os.path.exists(os.path.join(tmpdir, 'ts')):
+                pass
+            # Wait till docker builds image and starts up
+            # Sleep to wait for python http server to start
+            time.sleep(20)
+            http = urllib3.PoolManager()
```

Comment:
We already have requests library available. Is there a reason You're using urllib3 directly instead of using requests?

Reply:
Both requests and urllib3 are not used anywhere in the project and not specified in the setup.py requirements.

Reply:
Ah, I think they're brought in as dependencies of some other library we use (docker maybe?)